### ISSUE TITLE
Exception in thread Thread- OSError: [Errno 3] No such process

### ISSUE THREAD

**Issue by harelassayag** [scylladb/scylla-cluster-tests#530]
Scylla version:  2018.1.rc0-0.20180401.4a42fc87f
happend while streaming during Decommission nemesis
```
2018-04-03 10:03:48,189 nemesis          L0189 DEBUG| sdcm.nemesis.ChaosMonkey: Set current_disruption -> Decommission Node longevity-tls-50gb-4d-ent-2018-1-db-node-5d7b4176-0-13 [35.231.102.138 | 10.142.0.6] (s
....
2018-04-03 10:26:06,979 wait             L0044 DEBUG| Node longevity-tls-50gb-4d-ent-2018-1-db-node-5d7b4176-0-15 [35.196.20.110 | 10.142.0.6] (seed: None): Waiting for DB services to be up (241.587319136 s)
2018-04-03 10:26:07,503 remote           L0197 DEBUG| [10.142.0.6] [stdout] Apr 03 10:26:07 info   |  [shard 0] stream_session - [Stream #6812ced0-3729-11e8-bf63-000000000004] Streaming plan for Bootstrap-keyspace1-index-10 succeeded, peers={10.142.0.5}, tx=0 KiB, 0.00 KiB/s, rx=12802 KiB, 2103.52 KiB/s
2018-04-03 10:26:07,503 remote           L0197 DEBUG| [10.142.0.6] [stdout] Apr 03 10:26:07 info   |  [shard 0] range_streamer - Bootstrap with keyspace1 for keyspace=10.142.0.5 succeeded, took 219 seconds
2018-04-03 10:26:07,505 remote           L0197 DEBUG| [10.142.0.5] [stdout] Apr 03 10:26:07 info   |  [shard 8] stream_session - [Stream #6812ced0-3729-11e8-bf63-000000000004] Streaming plan for Bootstrap-keyspace1-index-10 succeeded, peers={10.142.0.6}, tx=12802 KiB, 2102.14 KiB/s, rx=0 KiB, 0.00 KiB/s
2018-04-03 10:26:11,473 output           L0642 DEBUG| Exception in thread Thread-25783:
2018-04-03 10:26:11,474 output           L0642 DEBUG| Traceback (most recent call last):
2018-04-03 10:26:11,474 output           L0642 DEBUG|   File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
2018-04-03 10:26:11,474 output           L0642 DEBUG|     self.run()
2018-04-03 10:26:11,474 output           L0642 DEBUG|   File "/usr/lib64/python2.7/threading.py", line 1084, in run
2018-04-03 10:26:11,474 output           L0642 DEBUG|     self.function(*self.args, **self.kwargs)
2018-04-03 10:26:11,474 output           L0642 DEBUG|   File "/usr/lib/python2.7/site-packages/avocado/utils/process.py", line 507, in send_signal
2018-04-03 10:26:11,474 output           L0642 DEBUG|     self._popen.send_signal(sig)
2018-04-03 10:26:11,475 output           L0642 DEBUG|   File "/usr/lib64/python2.7/subprocess.py", line 1546, in send_signal
2018-04-03 10:26:11,475 output           L0642 DEBUG|     os.kill(self.pid, sig)
2018-04-03 10:26:11,475 output           L0642 DEBUG| OSError: [Errno 3] No such process
2018-04-03 10:26:11,475 output           L0642 DEBUG| 
2018-04-03 10:26:12,873 remote           L0197 DEBUG| [10.142.0.6] [stdout] Apr 03 10:26:12 info   |  [shard 0] stream_session - [Stream #68131cf0-3729-11e8-bf63-000000000004] Streaming plan for Bootstrap-keyspace1-index-10 succeeded, peers={10.142.0.43}, tx=0 KiB, 0.00 KiB/s, rx=237258 KiB, 20694.16 KiB/s
2018-04-03 10:26:12,874 remote           L0197 DEBUG| [10.142.0.6] [stdout] Apr 03 10:26:12 info   |  [shard 0] range_streamer - Bootstrap with keyspace1 for keyspace=10.142.0.43 succeeded, took 224 seconds
2018-04-03 10:26:12,874 remote           L0197 DEBUG| [10.142.0.6] [stdout] Apr 03 10:26:12 info   |  [shard 0] range_streamer - Bootstrap with 10.142.0.5 for keyspace=mview, 19 out of 191 ranges: ranges = 19
2018-04-03 10:26:12,874 remote           L0197 DEBUG| [10.142.0.6] [stdout] Apr 03 10:26:12 info   |  [shard 0] stream_session - [Stream #6ee8af90-3729-11e8-bf63-000000000004] Executing streaming plan for Bootstrap-mview-index-0 with peers={10.142.0.5}, master
```
the test is running normally after that load has continued and further disruption was ran:
`2018-04-03 10:37:47,022 nemesis          L0189 DEBUG| sdcm.nemesis.ChaosMonkey: Set current_disruption -> StopWaitStartService Node longevity-tls-50gb-4d-ent-2018-1-db-node-5d7b4176-0-15 [35.196.20.110 | 10.142.0.6] (seed: False)
`

**Comment by roydahan** [comment]
@bentsi , can you please take a look?

**Comment by bentsi** [comment]
@roydahan sure

**Comment by bentsi** [comment]
Looking in the avocado stderr we can see that SSH connection was closed due to broken pipe.
This happened after the decommission removed the node.
```Connection to 10.142.0.6 closed.
packet_write_wait: Connection to 10.142.0.6 port 22: Broken pipe
Exception in thread Thread-25783:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 1084, in run
    self.function(*self.args, **self.kwargs)
  File "/usr/lib/python2.7/site-packages/avocado/utils/process.py", line 507, in send_signal
    self._popen.send_signal(sig)
  File "/usr/lib64/python2.7/subprocess.py", line 1546, in send_signal
    os.kill(self.pid, sig)
OSError: [Errno 3] No such process
```
The bug is in the implementation of destroy method of all classes that derive from BaseNode.
Currently the sequence of actions on destroy was as follows:

1. terminate the node instance using appropriate api (AWS, GCE, etc)
2. delete Remote obj (that triggers closing the SSH connection)
3. stop_task_threads (_backtrace_thread, _journal_thread)


The problem in this sequence is that SSH connection should be closed when all its users are exited.
The correct sequence is:

1. stop_task_threads (_backtrace_thread, _journal_thread)
2. delete Remote obj (that triggers closing the SSH connection)
3. terminate the node instance using appropriate api (AWS, GCE, etc)

In addition while closing the connection, if the SSH subprocess is already killed
for some reason, OSError exception is thrown and should be handled. Currently decided just to warn to log when this this situation happens.

**Comment by roydahan** [comment]
cc @amoskong

**Comment by bentsi** [comment]
@roydahan you can close the issue. It was fixed by #533
